tests/run/i3634.scala

```diff
@@ -0,0 +1,11 @@
+trait T {
+  case object Foo
+}
+
+object Bar extends T
+
+object Test {
+  def main(args: Array[String]): Unit = {
+    assert(Bar.Foo == Bar.Foo) // false
```

Comment:
I would use `eq` instead of `==` to be extra sure that nothing funny is going on.

Reply:
Also add a `object Baz extends T` and `assert(Bar.Foo ne Baz.Foo)`